### PR TITLE
[batch] SQL and Interface improvements

### PR DESCRIPTION
**Pull request by jigold**
@cseed I have a question about batch state when n_jobs == 0. Should this be `ready`? Should `complete` be True or False? I think this will be obsolete once all jobs are submitted with a batch.

- Added triggers to update the Batch state on Job insert, state update
- Got rid of delete and cancel on Job and list_jobs
- Added foreign keys to database with delete cascade
- Misc naming changes for clarity: get_record -> get_records
- Changed list_batches to not include individual jobs but batches/N still returns the jobs

### CODE REVIEW

**Comment by jigold** [comment]
Closing for now.

**Comment by jigold** [comment]
@cseed Here's the PR you requested.

**Comment by jigold** [comment]
Stacked on #6015

**Comment by cseed** [comment]
I merged your other PR.  Can you rebase this one?

**Comment by jigold** [comment]
Done

File: batch/batch/server/database.py [hunk 1]
```diff
@@ -25,7 +25,18 @@ def __init__(self, db):
...
-        return await super().get_record({'id': ids}, fields)
+        return await super().get_records({'id': ids}, fields)
+
+    async def get_undeleted_records(self, ids, user):
+        async with self._db.pool.acquire() as conn:
+            async with conn.cursor() as cursor:
+                batch_name = self._db.batch.name
+                where_template, where_values = make_where_statement({'id': ids, 'user': user})
+                sql = f"""SELECT * FROM `{self.name}` WHERE {where_template} AND EXISTS 
+                (SELECT id from `{batch_name}` WHERE `{batch_name}`.id = batch_id AND `{batch_name}`.deleted = 0)"""
```
Comment (cseed) [inline batch/batch/server/database.py]: Do you want `.deleted = FALSE`?   It seems we're using `BOOLEAN` and `TRUE` elsewhere.  Or just `.deleted` with no comparison, integers appear to get implicitly casted to boolean values as needed.

File: batch/batch/server/database.py [hunk 2]
```diff
@@ -119,7 +130,19 @@ def __init__(self, db):
...
+    async def get_records_where(self, condition):
+        return await super().get_records(condition)
 
     async def has_record(self, id):
         return await super().has_record({'id': id})
+
+    async def purge_finished_deleted_records(self):
+        async with self._db.pool.acquire() as conn:
+            async with conn.cursor() as cursor:
+                sql = f"DELETE FROM `{self.name}` WHERE `deleted` = TRUE AND `n_completed` = `n_jobs`"
```
Comment (cseed) [inline batch/batch/server/database.py]: Could this batch still be open?  What's the relationship between deleted and open?

File: batch/batch/server/server.py [hunk 3]
```diff
@@ -726,58 +701,58 @@ def wrapped(request, *args, **kwargs):
...
+        if record is not None and not record['deleted']:
             attributes = json.loads(record['attributes'])
             userdata = json.loads(record['userdata'])
+
+            print("\n")
+            print("is_open", record['is_open'])
+            print("n_completed", record['n_completed'])
+            print("n_jobs", record['n_jobs'])
+            print("n_failed", record['n_failed'])
+            print('n_cancelled', record['n_cancelled'])
```
Comment (cseed) [inline batch/batch/server/server.py]: Delete.

File: batch/batch/server/server.py [hunk 4]
```diff
@@ -451,13 +455,12 @@ def __init__(self, id, batch_id, attributes, callback, userdata, user, always_ru
...
+        # Can only be called via Batch.delete()
+
         children = [Job.from_record(record) for record in await db.jobs.get_children(self.id)]
         for child in children:
-            await child.cancel()
-
-        await db.jobs.delete_record(self.id)
-        await db.jobs_parents.delete_records_where({'job_id': self.id})
-        await db.jobs_parents.delete_records_where({'parent_id': self.id})
+            await child.delete()
```
Comment (cseed) [inline batch/batch/server/server.py]: This seems unnecessary.  You delete the all the jobs in Batch.delete.

File: batch/batch/server/server.py [hunk 5]
```diff
@@ -791,35 +766,45 @@ def from_record(record):
...
     async def delete(self):
+        # Batch is deleted from db in polling loop once all jobs are complete
+        await db.batch.update_record(self.id,
+                                     deleted=True,
+                                     is_open=False)
         jobs = await self.get_jobs()
         for j in jobs:
-            assert j.batch_id == self.id
-            await db.jobs.update_record(j.id, batch_id=None)
+            await j.delete()
```
Comment (cseed) [inline batch/batch/server/server.py]: I think we can remove Job.delete and just cancel here.  The delete cleanup logic will apply after everything winds down.

File: batch/batch/server/server.py [hunk 6]
```diff
@@ -791,35 +766,45 @@ def from_record(record):
...
+        await self.close()
         jobs = await self.get_jobs()
         for j in jobs:
             await j.cancel()
 
     async def delete(self):
+        # Batch is deleted from db in polling loop once all jobs are complete
+        await db.batch.update_record(self.id,
+                                     deleted=True,
+                                     is_open=False)
```
Comment (cseed) [inline batch/batch/server/server.py]: Should you set `self.is_open = False` and `self.delete = True`?

File: batch/batch/server/server.py [hunk 7]
```diff
@@ -876,7 +862,12 @@ def close_ttl(self):
     params = request.query
     user = userdata['ksa_name']
 
-    batches = [Batch.from_record(record) for record in await db.batch.get_record({'user': user})]
+    batches = []
+    for record in await db.batch.get_records_where({'user': user}):
+        batch = Batch.from_record(record)
+        if batch is not None:
```
Comment (cseed) [inline batch/batch/server/server.py]: How can batch be None here?

File: batch/batch/server/database.py (hunk 2 above)
Comment (cseed) [reply batch/batch/server/database.py]: I see below, delete closes.

File: batch/batch/server/server.py [hunk 8]
```diff
@@ -451,16 +455,8 @@ def __init__(self, id, batch_id, attributes, callback, userdata, user, always_ru
             await self._delete_k8s_resources()
 
     async def delete(self):
-        children = [Job.from_record(record) for record in await db.jobs.get_children(self.id)]
```
Comment (cseed) [inline batch/batch/server/server.py]: This is still wrong.  You can't set something to cancelled if it isn't ready.  I don't think we should have Job.delete, just call cancel.

File: batch/batch/server/server.py (hunk 7 above)
Comment (jigold) [reply batch/batch/server/server.py]: If the batch has been set to `deleted` in the database but the record still exists. I check for this in `Batch.from_record`

File: batch/batch/server/server.py (hunk 5 above)
Comment (cseed) [reply batch/batch/server/server.py]: I think this should be a call to cancel, see above comment.

Comment (jigold) [reply batch/batch/server/server.py]: The difference between delete and cancel is the log files are deleted in `delete` but not in `cancel`.

File: batch/batch/server/server.py (hunk 7 above)
Comment (cseed) [reply batch/batch/server/server.py]: You should push this into the database query.  Basically, no functions should return deleted rows now.

File: batch/batch/server/server.py [hunk 9]
```diff
@@ -791,35 +752,48 @@ def from_record(record):
...
+        await self.close()
         jobs = await self.get_jobs()
         for j in jobs:
             await j.cancel()
 
     async def delete(self):
+        # Batch is deleted from db in polling loop once all jobs are complete
+        self.is_open = False
+        self.deleted = True
         jobs = await self.get_jobs()
```
Comment (cseed) [inline batch/batch/server/server.py]: Getting the jobs between setting the local fields and the database fields worries me.

I also think we should set the local fields AFTER the database fields.  If the database write fails, they will be out of sync.

File: batch/batch/server/server.py [hunk 10]
```diff
@@ -726,58 +694,51 @@ def wrapped(request, *args, **kwargs):
...
-    return jsonify({})
-
-
 class Batch:
     MAX_TTL = 30 * 60
 
     @staticmethod
     def from_record(record):
-        if record is not None:
+        if record is not None and not record['deleted']:
```
Comment (cseed) [inline batch/batch/server/server.py]: I think this should assert not deleted.

File: batch/batch/server/server.py (hunk 9 above)
Comment (jigold) [reply batch/batch/server/server.py]: My concern was if the purge of records happens after the batch record in the database is set to deleted but before we get the job records out.

File: batch/batch/server/server.py (hunk 8 above)
Comment (jigold) [reply batch/batch/server/server.py]: What about the log files?

File: batch/batch/server/server.py (hunk 9 above)
Comment (cseed) [reply batch/batch/server/server.py]: Oh, good point.

What about just calling self.cancel() and then set deleted=True in the db.

File: batch/batch/server/server.py [hunk 11]
```diff
@@ -791,35 +745,54 @@ def from_record(record):
...
+        await self.close()
         jobs = await self.get_jobs()
         for j in jobs:
             await j.cancel()
 
+    async def mark_deleted(self):
+        await self.close()
+        await db.batch.update_record(self.id,
+                                     deleted=True)
+        self.is_open = False
```
Comment (cseed) [inline batch/batch/server/server.py]: This is already done by close.  This code is duplicating the logic in cancel, and I think can just be:

```
  self.cancel()
  await db.batch.update_record(self.id, deleted=True)
  self.deleted = True
```

File: batch/batch/server/server.py [hunk 12]
```diff
@@ -259,6 +259,11 @@ def _has_next_task(self):
         uri = await write_gs_log_file(app['blocking_pool'], INSTANCE_ID, self.id, task_name, log)
         await db.jobs.update_log_uri(self.id, task_name, uri)
 
+    async def _delete_logs(self):
+        for idx, jt in enumerate(self._tasks):
+            if idx < self._task_idx:
+                await delete_gs_log_file(app['blocking_pool'], INSTANCE_ID, self.id, jt.name)
```
Comment (cseed) [inline batch/batch/server/server.py]: I'm worried about what happens when this fails.  Say we've deleted several of the logs, and a later one times out.  That will throw out of the cleanup loop, and we'll try again later.  When we go to delete the first logs again, we'll get a 404 ... throwing out of the cleanup loop.

In other words, I think this should ignore 404.  Then I think I'm happy.

File: batch/batch/server/server.py [hunk 13]
```diff
@@ -1111,10 +1088,22 @@ def __anext__(self):
         await asyncio.sleep(REFRESH_INTERVAL_IN_SECONDS)
 
 
+async def db_cleanup_event_loop():
+    await asyncio.sleep(60)
+    while True:
+        try:
+            batches = [Batch.from_record(record) for record in await db.batch.get_finished_deleted_records()]
+            await asyncio.gather(*[batch.delete() for batch in batches])
```
Comment (cseed) [inline batch/batch/server/server.py]: This is a  low priority cleanup thread.  No need to run these delete operations concurrently.  Just loop over the records and call delete on the batches one by one.

**Comment by jigold** [comment]
The ci page says this build was successful. Not sure why it is not showing up in GitHub.

**Comment by cseed** [comment]
It's a successful build for a previous version of master.  I'm working on a PR to note when ci2 builds are out of date.  We only update on approved PR at a time, so it is in a queue.

**Comment by cseed** [comment]
It is rebuilding now.

**Comment by danking** [comment]
There's some change to set ordering in python https://ci2.hail.is/jobs/26259/log.

**Comment by jigold** [comment]
I had to fix the tests to test equality rather than the expected set itself.
https://docs.python.org/3.5/library/doctest.html#warnings

**Pull request merged**